drivers/virtio/virtio_mmio.c

```diff
@@ -0,0 +1,419 @@
+/*
+ * Copyright (c) 2025 TOKITA Hiroshi
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/kernel/mm.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/spinlock.h>
+#include <zephyr/sys/barrier.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/virtio/virtio.h>
+#include <zephyr/virtio/virtqueue.h>
+#include "assert.h"
+
+#define DT_DRV_COMPAT virtio_mmio
+
+LOG_MODULE_REGISTER(virtio_mmio, CONFIG_VIRTIO_LOG_LEVEL);
+
+/*
+ * Based on Virtual I/O Device (VIRTIO) Version 1.3 specification:
+ * https://docs.oasis-open.org/virtio/virtio/v1.3/csd01/virtio-v1.3-csd01.pdf
+ */
+
+#define VIRTIO_MMIO_MAGIC_VALUE             0x000
+#define VIRTIO_MMIO_VERSION                 0x004
+#define VIRTIO_MMIO_DEVICE_ID               0x008
+#define VIRTIO_MMIO_VENDOR_ID               0x00c
+#define VIRTIO_MMIO_DEVICE_FEATURES         0x010
+#define VIRTIO_MMIO_DEVICE_FEATURES_SEL     0x014
+#define VIRTIO_MMIO_DRIVER_FEATURES         0x020
+#define VIRTIO_MMIO_DRIVER_FEATURES_SEL     0x024
+#define VIRTIO_MMIO_QUEUE_SEL               0x030
+#define VIRTIO_MMIO_QUEUE_SIZE_MAX          0x034
+#define VIRTIO_MMIO_QUEUE_SIZE              0x038
+#define VIRTIO_MMIO_QUEUE_READY             0x044
+#define VIRTIO_MMIO_QUEUE_NOTIFY            0x050
+#define VIRTIO_MMIO_INTERRUPT_STATUS        0x060
+#define VIRTIO_MMIO_INTERRUPT_ACK           0x064
+#define VIRTIO_MMIO_STATUS                  0x070
+#define VIRTIO_MMIO_QUEUE_DESC_LOW          0x080
+#define VIRTIO_MMIO_QUEUE_DESC_HIGH         0x084
+#define VIRTIO_MMIO_QUEUE_DRIVER_LOW        0x090
+#define VIRTIO_MMIO_QUEUE_DRIVER_HIGH       0x094
+#define VIRTIO_MMIO_QUEUE_DEVICE_LOW        0x0a0
+#define VIRTIO_MMIO_QUEUE_DEVICE_HIGH       0x0a4
+#define VIRTIO_MMIO_QUEUE_SHM_SEL           0x0ac
+#define VIRTIO_MMIO_QUEUE_SHM_LEN_LOW       0x0b0
+#define VIRTIO_MMIO_QUEUE_SHM_LEN_HIGH      0x0b4
+#define VIRTIO_MMIO_QUEUE_SHM_BASE_LOW      0x0b8
+#define VIRTIO_MMIO_QUEUE_SHM_BASE_HIGH     0x0bc
+#define VIRTIO_MMIO_QUEUE_RESET             0x0c0
+#define VIRTIO_MMIO_QUEUE_CONFIG_GENERATION 0x0fc
+#define VIRTIO_MMIO_CONFIG                  0x100
+
+#define DEVICE_STATUS_ACKNOWLEDGE 0
+#define DEVICE_STATUS_DRIVER      1
+#define DEVICE_STATUS_DRIVER_OK   2
+#define DEVICE_STATUS_FEATURES_OK 3
+#define DEVICE_STATUS_NEEDS_RESET 6
+#define DEVICE_STATUS_FAILED      7
+
+#define VIRTIO_F_VERSION_1 32
+
+/* Ranges of feature bits for specific device types (see spec 2.2)*/
+#define DEV_TYPE_FEAT_RANGE_0_BEGIN 0
+#define DEV_TYPE_FEAT_RANGE_0_END   23
+#define DEV_TYPE_FEAT_RANGE_1_BEGIN 50
+#define DEV_TYPE_FEAT_RANGE_1_END   127
+
+#define VIRTIO_MMIO_QUEUE_INTERRUPT                1
+#define VIRTIO_MMIO_DEVICE_CONFIGURATION_INTERRUPT 2
```

Comment:
These defs are now provided by `virito_common.h`.

Reply:
Fixed to use common definition.